### DISCUSSION TITLE
How does `vmap` work when using non-vectorized XLA functions?

### DISCUSSION THREAD

**Discussion by joelberkeley** [jax-ml/jax#11686] [Q&A]
I'm trying to understand how `vmap` works.

According to [the docs](https://jax.readthedocs.io/en/latest/jax-101/03-vectorization.html?highlight=vmap#automatic-vectorization), `vmap` vectorizes functions under the hood. I assume it does this by arranging axes appropriately such that it can use the vectorization already available in XLA. For example, `xla::Cholesky` allows shape `leading ++ [n, n]` so vectorising cholesky application over an `[n, n]` matrix is trivial - we can just pass it into `xla::Cholesky` and it will handle the leading dims. Is that correct?

If that is correct, then how does jax handle XLA functions that _don't_ allow vectorization, such as the predicate in `xla::Cond`. How does this work, for example?
```
>>> jax.vmap(lambda x: jax.lax.cond(x, lambda: 1, lambda: 0))(np.array([True, False]))
DeviceArray([1, 0], dtype=int32, weak_type=True)
```

**Comment by soraros** [comment]
`vmap` of `lax.cond` is lowered into `select`, as can be seem in `_cond_batching_rule`.
https://github.com/google/jax/blob/51120069037674fcd9484b9b68c94869d325973c/jax/_src/lax/control_flow/conditionals.py#L308-L312

**Reply by joelberkeley** [reply]
Thanks, and what about other functions whose arguments are restricted in their dimensionality, like the key in PRNG, and HOF closures? I tried using the mapped-over argument in the closure in `reduce`, and it failed
```python
>>> jax.vmap(lambda x: jax.lax.reduce(np.array([1, 2, 3]), 0, (lambda y, z: y + z + x), 0))(np.array([0, 1]))
...
NotImplementedError: Reduction computations can't close over Tracers. Please open an issue at https://github.com/google/jax.
```
Is that because it's not allowed to use the thing you're mapping over in the closure in jax? (note that closure may well not obey the monoid laws required for `xla::Reduce`, but I don't believe that matters for this discussion).

I notice that jax doesn't expose the computation in `xla::Sort` and I don't know if `xla::Map` is used at all in jax given the existence of vmap. I notice it does work for `xla::Cond` but I'm guessing that's because it defers to `xla::Select`
```python
>>> jax.vmap(lambda x: jax.lax.cond(True, lambda: x, lambda: 0))(np.array([1, 2]))
DeviceArray([1, 2], dtype=int32, weak_type=True)
```

**Reply by joelberkeley** [reply]
another function that comes to mind that I haven't tried is `xla::While` - there's a scalar predicate in that

**Reply by joelberkeley** [reply]
and another: indexing with dynamic indices
```python
>>> jax.vmap(lambda x: jnp.array([3, 4, 5])[x])(jnp.array([0, 2]))
DeviceArray([3, 5], dtype=int32)
```
I imagine that's more complicated than the others because iiuc XLA doesn't support indexing in all the ways numpy does. Indeed we see another error for a much less pleasant
```python
>>> jax.vmap(lambda x: jnp.array([3, 4, 5])[:x])(jnp.array([2]))
```
